src/base/bittorrent/customstorage.cpp

```diff
@@ -35,6 +35,191 @@
 #include "base/utils/fs.h"
 #include "common.h"
 
+#if (LIBTORRENT_VERSION_NUM >= 20000)
+#include <libtorrent/session.hpp>
+
+std::unique_ptr<lt::disk_interface> customDiskIOConstructor(
+        lt::io_context &ioContext, const lt::settings_interface &settings, lt::counters &counters)
+{
+    return std::make_unique<customDiskIOThread>(lt::default_disk_io_constructor(ioContext, settings, counters));
+}
+
+customDiskIOThread::customDiskIOThread(std::unique_ptr<libtorrent::disk_interface> nativeDiskIOThread)
+    : m_nativeDiskIO {std::move(nativeDiskIOThread)}
+{
+}
+
+lt::storage_holder customDiskIOThread::new_torrent(const lt::storage_params &storageParams, const std::shared_ptr<void> &torrent)
+{
+    lt::storage_holder storageHolder = m_nativeDiskIO->new_torrent(storageParams, torrent);
+
+    const QString savePath = Utils::Fs::expandPathAbs(QString::fromStdString(storageParams.path));
+    m_storageData[static_cast<lt::storage_index_t>(storageHolder)] = {
+            savePath
+            , storageParams.mapped_files ? *storageParams.mapped_files : storageParams.files
+            , storageParams.priorities};
+
+    return storageHolder;
+}
+
+void customDiskIOThread::remove_torrent(lt::storage_index_t storage)
+{
+    m_nativeDiskIO->remove_torrent(storage);
```

Comment:
The new added code is quite long (compared to previous customStorage code) and most of them are boilerplate/trivial cases.
Do you think libtorrent should provide a base implementation so that we can ignore the trivial cases?

Reply:
The problem is that the existing implementations are declared final in libtorrent-2.0, so I can't inherit it like I did before, and I have to use a wrapper. For some reason, @arvidn decided that it doesn't make sense to extend existing implementations.

Reply:
not only is it final, the type itself is not exposed as part for the public interface. The reason for this is that doing so sets much of its implementation in ABI-stone. I think it's too high a cost to pay to not be able to improve (and sometimes fix bugs) because it would alter the ABI. As it is now, both disk I/O implementation are only exposed via the `disk_interface`.

Reply:
I'm not really experienced in ABI compatibility issues and the all the class hierarchy.
Currently, I could image libtorrent provides a reference struct `lt::example_disk_interface` (which is derived from `lt::disk_interface`) in header-form. The `lt::example_disk_interface` would contain all boilerplate code that are seen here and users of the library/header could derive from it and only need to overwrite some of the methods when using it, would that be feasible?

Reply:
it could be, however, I don't really think this is what you *want*. I think it's done this way now because it's simple, and there aren't any other options really. Ideally these disk operations would be run in the disk thread, where they won't block the network thread, but also where they are guaranteed to not conflict with any other disk threads accessing the same files (this is an issue with this patch).

There is no customization point for this, but maybe one could be developed. Would it be sufficient to be able to rewrite file paths?

e.g. whenever libtorrent wants to open a file, you get an opportunity to rewrite the path it opens (and maybe do other things to the filesystem). Such customization point would likely also need to know what the current operation on the file is. I'll take another pass over this patch to understand what you're doing.

Reply:
>it could be, however, I don't really think this is what you want. I think it's done this way now because it's simple, and there aren't any other options really. 

You're probably right, I'm merely focusing on cutting the boilerplate in our current situation. I'm looking forward that there will be positive outcome/design that will benefit all of us in the end.

Reply:
>not only is it final, the type itself is not exposed as part for the public interface. The reason for this is that doing so sets much of its implementation in ABI-stone. I think it's too high a cost to pay to not be able to improve (and sometimes fix bugs) because it would alter the ABI.

If you care about ABI, you might consider PIMPL. Then default disk_interface implementation will only export the public interface, which will keep ABI stable. Or am I mistaken?

Reply:
right. the suggestion for libtorrent to provide a class that forwards to an interface is essentially pimp.